### PR TITLE
Fixing case in which `Trainer` hung while saving model in distributed training

### PR DESCRIPTION
**Pull request by TevenLeScao**
As found thanks to the great @mfuntowicz , the call to `store_flos` in `Trainer` can hang indefinitely, as it was only executed in the main thread and in some cases the other threads were already past this point. This PR moves this call in order to avoid this behaviour.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -740,6 +740,8 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
                     epoch_pbar.update(1)
                     continue
 
+                # print(inputs)
+
```
Comment (LysandreJik) [inline src/transformers/trainer.py:744]: Can you remove these two lines?

**Review by LysandreJik — APPROVED** [review]
Great, LGTM :)

File: src/transformers/trainer.py (hunk 1 above)
Comment (TevenLeScao) [reply src/transformers/trainer.py]: Wooooops it was a long debugging session

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

**Comment by julien-c** [comment]
Reminder that we need to add some CI infra (and tests) for multi-gpu and/or multi-node setups

**Pull request merged**